homeassistant/components/qnap/__init__.py

```diff
@@ -1 +1,85 @@
 """The qnap component."""
+from datetime import timedelta
+import logging
+
+from qnapstats import QNAPStats
+
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DEFAULT_PORT, DEFAULT_TIMEOUT, DOMAIN, PLATFORMS
+
+UPDATE_INTERVAL = timedelta(minutes=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass, config):
+    """Set up the qnap environment."""
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set the config entry up."""
+    host = config_entry.data[CONF_HOST]
+    protocol = "https" if config_entry.data.get(CONF_SSL) else "http"
+    api = QNAPStats(
+        host=f"{protocol}://{host}",
+        port=config_entry.data.get(CONF_PORT, DEFAULT_PORT),
+        username=config_entry.data[CONF_USERNAME],
+        password=config_entry.data[CONF_PASSWORD],
+        verify_ssl=config_entry.data.get(CONF_VERIFY_SSL),
+        timeout=DEFAULT_TIMEOUT,
+    )
+
+    async def async_update_data():
+        datas = {}
+        datas["system_stats"] = await hass.async_add_executor_job(api.get_system_stats)
+        datas["system_health"] = await hass.async_add_executor_job(
+            api.get_system_health
+        )
+        datas["smart_drive_health"] = await hass.async_add_executor_job(
+            api.get_smart_disk_health
+        )
+        datas["volumes"] = await hass.async_add_executor_job(api.get_volumes)
+        datas["bandwidth"] = await hass.async_add_executor_job(api.get_bandwidth)
+        return datas
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=UPDATE_INTERVAL,
+    )
```

Comment:
Adding a coordinator should really be done in a preliminary PR.
It would also be better to make a standalone class in `coordinator.py`, and use a dataclass instead of a dict for the data.

Reply:
Hmmm, never quite done this before 😬

Reply:
Same again, it is a very useful change but it isn't really related to config flow. You can add a coordinator as a preliminary PR (create a class in coordinator.py, and initialise it inside setup_platform) or as a follow-up PR (create a class in coordinator.py, and initialise it inside `async_setup_entry`)
Either way it shouldn't be in the same PR as the config flow.

Reply:
I've moved the coordinator out to it's own class. I'm thinking of submitting this in a follow up PR just to get this one merged. Is that ok?

See: https://github.com/disforw/qnap/commit/726f6f76f14cafa7e2dd18ec1ccc86532795709a